_P248_TempHumidity_AHT1x.ino

```diff
@@ -0,0 +1,339 @@
+
+//#ifdef PLUGIN_BUILD_TESTING
+
+#define PLUGIN_248
+#define PLUGIN_ID_248         248
+#define PLUGIN_NAME_248       "Environment - AHT10 [TESTING]"
+#define PLUGIN_VALUENAME1_248 "Temperature"
+#define PLUGIN_VALUENAME2_248 "Humidity"
+#define PLUGIN_VALUENAME3_248 "AbsoluteHumidity"
+#define PLUGIN_VALUENAME4_248 "DewPoint" // max 4 variables allowed
+
+
+#define AHT10_ADDRESS_0X38         0x38  //chip I2C address no.1 for AHT10/AHT15/AHT20, address pin connected to GND
+#define AHT10_ADDRESS_0X39         0x39  //chip I2C address no.2 for AHT10 only, address pin connected to Vcc
+
+#define AHT10_INIT_CMD             0xE1  //initialization command for AHT10/AHT15
+#define AHT20_INIT_CMD             0xBE  //initialization command for AHT20
+#define AHT10_START_MEASURMENT_CMD 0xAC  //start measurment command
+#define AHT10_NORMAL_CMD           0xA8  //normal cycle mode command, no info in datasheet!!!
+#define AHT10_SOFT_RESET_CMD       0xBA  //soft reset command
+
+#define AHT10_INIT_NORMAL_MODE     0x00  //enable normal mode
+#define AHT10_INIT_CYCLE_MODE      0x20  //enable cycle mode
+#define AHT10_INIT_CMD_MODE        0x40  //enable command mode
+#define AHT10_INIT_CAL_ENABLE      0x08  //load factory calibration coeff
+
+
+#define AHT10_DATA_MEASURMENT_CMD  0x33  //no info in datasheet!!! my guess it is DAC resolution, saw someone send 0x00 instead
+#define AHT10_DATA_NOP             0x00  //no info in datasheet!!!
+
+
+#define AHT10_MEASURMENT_DELAY     80    //at least 75 milliseconds
+#define AHT10_POWER_ON_DELAY       40    //at least 20..40 milliseconds
+#define AHT10_CMD_DELAY            350   //at least 300 milliseconds, no info in datasheet!!!
+#define AHT10_SOFT_RESET_DELAY     20    //less than 20 milliseconds
+
+#define AHT10_FORCE_READ_DATA      true  //force to read data
+#define AHT10_USE_READ_DATA        false //force to use data from previous read
+#define AHT10_ERROR                0xFF  //returns 255, if communication error is occurred
+
+ 
+#define AHTX0_I2CADDR_DEFAULT 0x38   ///< AHT default i2c address
+#define AHTX0_CMD_CALIBRATE 0xE1     ///< Calibration command
+#define AHTX0_CMD_TRIGGER 0xAC       ///< Trigger reading command
+#define AHTX0_CMD_SOFTRESET 0xBA     ///< Soft reset command
+#define AHTX0_STATUS_BUSY 0x80       ///< Status bit for busy
+#define AHTX0_STATUS_CALIBRATED 0x08 ///< Status bit for calibrated
+
+
+#ifdef USE_AHT2x
+  #define AHTX_CMD     0xB1 // Cmd for AHT2x
+#else
+  #define AHTX_CMD     0xE1 // Cmd for AHT1x
+#endif
+
+uint8_t AHTSetCalCmd[3]    = { AHTX_CMD, 0x08, 0x00 }; // load factory calibration coeff
+uint8_t AHTSetCycleCmd[3]  = { AHTX_CMD, 0x28, 0x00 }; // enable cycle mode
+uint8_t AHTMeasureCmd[3]   = { 0xAC, 0x33, 0x00 };     // start measurment command
+uint8_t AHTResetCmd        =   0xBA;                   // soft reset command
+
+uint8_t          AHT10_rawDataBuffer[6] = {AHT10_ERROR, 0, 0, 0, 0, 0};
```

Comment:
This makes that this plugin cannot be used with multiple instances of the same plugin.
Can you have a look at for example P107 in the main repository as that one does hardly keep anything in its P107_data_struct.h file and thus might be easy to see how to keep plugin specific data per task.

Reply:
actually this device needs to be alone on the I2C bus. Can we enforce the plugin to be added only once ?

Reply:
Why? It is an I2C device, so even if you cannot change its address, you can still use an I2C multiplexer.

Reply:
it is the datasheet: "3, Only a single AHT10 can be connected to the I 2 C bus and no other I 2 C 
devices can be connected ."
I have not test it with a multiplexer as I don't have one on hand.

Reply:
The whole idea of an I2C multiplexer is that it (temporary) creates a separate I2C bus.
So to the sensor it is a single I2C device if it is the only one present on the channel of the multiplexer.

But what bothers me is the sentence mentions "No other I2C devices" and not "No other AHT10 devices".
That does sound like it may not be I2C compliant.

Reply:
Hmm [Tasmota](https://tasmota.github.io/docs/AHT1x/) also has a similar warning...

Reply:
See also [this explanation](https://www.raspberrypi.org/forums/viewtopic.php?t=265734#p1624629)

So then please mark it in comments with big exclamation marks that this sensor is NOT I2C compliant so cannot be used along with other I2C sensors.

I think it may still be usable via an multiplexer, but we may need to test.... but to be honest with these red flags I wonder if we ever should include it in a build.
What makes this so special that given it is not compliant with such a common standard as I2C, that it is still worth considering to use it?

Reply:
ok.

Reply:
These are cheaper than the AHT20 which seems to be I2C compliant.
Most of my installed boards have only one i2c device on the bus thus as long these are stable should be good to be used in those situations. 
The BME280 sensor is more expensive. I have been using other temperature + humidity sensors myself but I had to solder them as no boards were available.